app/code/core/Mage/Core/Model/App.php

```diff
@@ -1393,28 +1425,30 @@ public function dispatchEvent($eventName, $args)
             }
             if ($events[$eventName] === false) {
                 continue;
-            } else {
-                $event = new Varien_Event($args);
-                $event->setName($eventName);
-                $observer = new Varien_Event_Observer();
             }
 
             foreach ($events[$eventName]['observers'] as $obsName => $obs) {
-                $observer->setData(['event' => $event]);
+                $observer = new Varien_Event_Observer([
+                    'event' => new Varien_Event([
+                        ...$obs['args'], // Default config.xml <args>
+                        ...$args,        // Mage::dispatchEvent() $args
+                        'name' => $eventName,
+                    ]),
+                    ...$obs['args'], // Default config.xml <args>
+                    ...$args,        // Mage::dispatchEvent() $args
+                ]);
                 Varien_Profiler::start('OBSERVER: ' . $obsName);
                 switch ($obs['type']) {
                     case 'disabled':
                         break;
                     case 'object':
                     case 'model':
                         $method = $obs['method'];
-                        $observer->addData($args);
```

Comment:
not sure if I'd remove this, wouldn't it break anybody's code that does $observer->getXXX() instead of $observer->getEvent()->getXXX()?

Reply:
It's added here:
https://github.com/MahoCommerce/maho/blob/977a78f2bca96608f074925b5f27e0c285cf6f96/app/code/core/Mage/Core/Model/App.php#L1438